src/ptz-gamepad.cpp

```diff
@@ -0,0 +1,416 @@
+	rightStickY = (abs(normRY) < DEADZONE_Y ? 0 : (abs(normRY) - DEADZONE_Y) * (normRY / abs(normRY)));
+
+	if constexpr (DEADZONE_X > 0.0f && DEADZONE_X < 1.0f)
+		rightStickX *= 1 / (1 - DEADZONE_X);
+	if constexpr (DEADZONE_Y > 0.0f && DEADZONE_Y < 1.0f)
+		rightStickY *= 1 / (1 - DEADZONE_Y);
+
+	if (lastRightStickX != rightStickX || lastRightStickY != rightStickY)
+	{
+		ptz_debug("PTZGamepad: right stick changed x=%.3f, y=%.3f", rightStickX, rightStickY);
+		emit rightAxisChanged(rightStickX, rightStickY);
+	}
+}
+
+void PTZGamePad::inputHandleLeftStick()
+{
+	const float lastLeftStickY = leftStickY;
+
+	const float normLY = fmaxf(-1, (float)state.Gamepad.sThumbLY / 32767);
+
+	leftStickY = (abs(normLY) < DEADZONE_Y ? 0 : (abs(normLY) - DEADZONE_Y) * (normLY / abs(normLY)));
+
+	if constexpr (DEADZONE_Y > 0.0f && DEADZONE_Y < 1.0f)
+		leftStickY *= 1 / (1 - DEADZONE_Y);
+
+	if (lastLeftStickY != leftStickY)
+	{
+		ptz_debug("PTZGamepad: left stick changed x=%.3f", leftStickY);
+		emit leftAxisChanged(leftStickY);
+	}
+}
+
+void PTZGamePad::inputHandleButtons()
+{
+	unsigned short buttons = 0;
+
+	for (int bIndex = 0; bIndex < BUTTONMAP_COUNT; ++bIndex)
+	{
+		const ButtonMapping* button = &buttonMap[bIndex];
+		const bool buttonPressedLast = ((buttonsDown & button->xinputButton) != 0);
+		const bool buttonPressedNow = ((state.Gamepad.wButtons & button->xinputButton) != 0);
+		if (!buttonPressedLast && buttonPressedNow)
+		{
+			buttons |= button->xinputButton;
+		}
+	}
+
+	buttonsDown = state.Gamepad.wButtons;
+
+	if (buttons != 0)
+	{
+		emit buttonDownChanged(buttons);
+	}
+}
+
+void PTZGamePad::uiGamepadStatus(byte status)
+{
+	if (!uiSettings)
+	{
+		return;
+	}
+	else if (status == GAMEPAD_STATUS_BLANK)
+	{
+		uiSettings->gamepadActiveText->setText(QStringLiteral(""));
+	}
+	else if (status == GAMEPAD_STATUS_NONE)
+	{
+		uiSettings->gamepadActiveText->setText(QStringLiteral("No Gamepad Detected"));
+	}
+	else if (status == GAMEPAD_STATUS_FOUND)
+	{
+		uiSettings->gamepadActiveText->setText(QStringLiteral("Found Gamepad"));
+	}
+}
+
+void PTZGamePad::setGamepadEnabled(bool enabled, Ui_PTZSettings *ui)
+{
+	if (ui)
+	{
+		uiSettings = ui;
+	}
+
+	if (enabled) 
+	{
+		initializeGamepads();
+		if (gamepadID != INVALID_GAMEPAD_ID) 
+		{
+			emit uiGamepadStatusChanged(GAMEPAD_STATUS_FOUND);
+			startThread();
+		}
+		else
+		{
+			emit uiGamepadStatusChanged(GAMEPAD_STATUS_NONE);
+			stopThread();
+		}
+	} 
+	else 
+	{
+		gamepadID = INVALID_GAMEPAD_ID;
+		emit uiGamepadStatusChanged(GAMEPAD_STATUS_BLANK);
+		stopThread();
+	}
+}
+
+void PTZGamePad::initializeGamepads()
+{
+	gamepadID = INVALID_GAMEPAD_ID;
+	for (DWORD i = 0; i < XUSER_MAX_COUNT; i++) 
+	{
+		ZeroMemory(&state, sizeof(XINPUT_STATE));
+
+		const DWORD dwResult = XInputGetState(i, &state);
+
+		if (dwResult == ERROR_SUCCESS) {
+
+			gamepadID = i;
+			break;
+		}
+	}
+}
+
+void PTZGamePad::setPanTilt(float stickX, float stickY)
+{
+	controls->setPanTilt(stickX, stickY);
+}
```

Comment:
I didn't copy QGamepad on purpose to make it my own. However, I will update this to follow a registration api in controls.

Reply:
😁 I understand wanting to make it your own. However from a maintenance perspective the closer it is to QGamepad the better. It is an established API that is well abstracted.

In fact, my preferred solution would be to import QGamepad and fix the bugs rather than writing something new. That would give a path for enabling support in MacOS and Linux. But, you're doing this work, not me. I'm not going to reject your gamepad support just based on that preference.

Reply:
It's much more similar in the newest revision.

Reply:
👍